src/inverse/ws.rs

```diff
@@ -0,0 +1,353 @@
+    // Order quantity
+    pub qty: &'a str,
+    // Millisecond timestamp
+    pub time: u64,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(untagged)]
+pub enum PublicResponse<'a> {
+    #[serde(borrow)]
+    OrderBookL2Snapshot(BaseResponseWithTimestamp<'a, Vec<OrderBookItem<'a>>>),
+    OrderBookL2Delta(BaseResponseWithTimestamp<'a, OrderBookDelta<'a>>),
+    Trade(BaseResponse<'a, Vec<Trade<'a>>>),
+    Insurance(BaseResponse<'a, Vec<Insurance<'a>>>),
+    InstrumentInfoSnapshot(BaseResponseWithTimestamp<'a, InstrumentInfoSnapshot<'a>>),
+    InstrumentInfoDelta(BaseResponseWithTimestamp<'a, InstrumentInfoDelta<'a>>),
+    Kline(BaseResponseWithTimestamp<'a, Vec<Kline>>),
+    Liquidation(BaseResponse<'a, Liquidation<'a>>),
+}
+
+pub struct PublicWebSocketApiClient {
+    pub uri: String,
+    subscriptions: Vec<String>,
+}
+
+#[derive(Serialize, Debug)]
+struct Subscription {
+    op: String,
+    args: Vec<String>,
+}
+
+impl PublicWebSocketApiClient {
+    pub fn new(uri: &str) -> Self {
+        return PublicWebSocketApiClient {
+            uri: uri.to_string(),
+            subscriptions: Vec::new(),
+        };
+    }
+
+    pub fn subscribe_order_book_l2_25(&mut self, symbols: &Vec<String>) {
+        self.subscribe("orderBookL2_25", symbols);
+    }
+
+    pub fn subscribe_order_book_l2_200(&mut self, symbols: &Vec<String>) {
+        self.subscribe("orderBook_200.100ms", symbols);
+    }
+
+    pub fn subscribe_trade(&mut self, symbols: &Vec<String>) {
+        self.subscribe("trade", symbols);
+    }
+
+    pub fn subscribe_insurance(&mut self, symbols: &Vec<String>) {
+        self.subscribe("insurance", symbols);
+    }
+
+    pub fn subscribe_instrument_info(&mut self, symbols: &Vec<String>) {
+        self.subscribe("instrument_info.100ms", symbols);
+    }
+
+    pub fn subscribe_kline(&mut self, symbols: &Vec<String>, interval: &str) {
+        let topic = format!("klineV2.{}", interval);
+        self.subscribe(&topic, symbols);
+    }
+
+    pub fn subscribe_liquidation(&mut self, symbols: &Vec<String>) {
+        self.subscribe("liquidation", symbols);
+    }
+
+    pub fn run<Callback: FnMut(PublicResponse)>(&self, mut callback: Callback) -> Result<()> {
+        let req = Url::parse(&self.uri).unwrap();
+        let (mut ws, _) = connect(req).expect("Can't connect");
+
+        let (tx, rx) = mpsc::channel::<String>();
+        spawn_ping_thread(tx);
+
+        for subscription in &self.subscriptions {
+            ws.write_message(Message::Text(subscription.clone()))
+                .unwrap();
+        }
+
+        loop {
+            if let Ok(ping) = rx.try_recv() {
+                ws.write_message(Message::Text(ping)).unwrap();
+            }
+
+            if let Ok(msg) = ws.read_message() {
+                match msg {
+                    Message::Text(content) => {
+                        if let Ok(_) = serde_json::from_str::<Pong>(&content) {
+                            continue;
+                        }
```

Comment:
I don't love this but it did clear a bunch of ERROR logs, for the cased where `Pong` messages weren't matched by `PublicResponse`.

Too bad line 325 doesn't match these messages :thinking:

Reply:
It's not a big problem, let's just make it work first and then take performance and code quality into account.

When I was verifying this PR, I found that instrument info snap response could not be deserialized and the response of perpetual is different from futures. I fixed them locally. I will merge this PR and push the patch to main branch.

Reply:
> let's just make it work first and then take performance and code quality into account

:rocket: 

> I fixed them locally. I will merge this PR and push the patch to main branch.

Thanks!